Update draft-ldbc-cats-framework.md (#26)
boucadair/can-framework: draft-ldbc-cats-framework.md

# Introduction

> Edge computing architectures have been expanding from single edge nodes to multiple, sometimes collaborative, edge nodes to address various issues, like long response times, or suboptimal service and network resource usage.

Reviewer: ```suggestion
Edge computing architectures have been expanding from single edge nodes to multiple, sometimes collaborative, edge nodes to address various issues (e.g., long response times or suboptimal service and network resource usage).
```

The underlying networking infrastructures that include edge computing resources usually provide relatively static service dispatching. In such infrastructures, service-specific traffic is often directed to the closest edge resource from a routing perspective without considering the actual network state (e.g., traffic congestion conditions).

As described in {{?I-D.yao-cats-ps-usecases}}, traffic steering that takes into account computing resource metrics would benefit several services, including latency-sensitive service like immersive services that rely upon the use of augmented reality or virtual reality (AR/VR) techniques. This document provides an architectural framework that aims at facilitating the making of compute- and network-aware traffic steering decisions in networking environments where edge computing resources are deployed.

> The Computing-Aware Traffic Steering (CATS) framework assumes that there may be multiple service instances running on different edge nodes, globally providing one given (connectivity) service. A single edge node may have limited computing resources available at a given time, whereas the various edge nodes may experience different resource availability issues over time. A single edge node may also host multiple instances of a service or just one service instance.

Reviewer: ```suggestion
The Computing-Aware Traffic Steering (CATS) framework assumes that there may be multiple service instances running on different edge nodes, globally providing one given service. A single edge node may have limited computing resources available at a given time, whereas the various edge nodes may experience different resource availability issues over time. A single edge node may also host multiple instances of a service or just one service instance.
```

Computing-Aware Traffic Steering (CATS):
 :  A traffic engineering approach that takes into account the dynamic nature of computing resources and network state to optimize service-specific traffic forwarding towards a given service instance. Various relevant metrics may be used to enforce such computing-aware traffic steering policies.

CATS Service ID (CS-ID):
 : An identifier representing a service, which the clients use to access it.

CATS Binding ID (CB-ID):
 : An identifier of a single service instance or location of a given service instance (CS-ID).

Service:
  : A monolithic function that is provided by a network operator according to a specification. A composite service can be built by orchestrating multiple monolithic services.

Service instance:
  : A run-time environment (e.g., a server or a process on a server) that makes a service instance available (i.e., up and running). One service can be accessed through multiple instances running at the same or different locations.

Service demand:
: The demand for a service identified by a CATS Service ID (CS-ID). See {{cats-ids}} for more details.

Service request:
 : The request for a specific service instance.

CATS-Router:

>  : A network device (usually located at the edge of the network) that makes forwarding decisions based on CATS information to steer traffic speciic to a  service demand towards a corresponding yet selected service instance. The selection of a service instance relies upon a multi-metric CATS-based path computation.

Reviewer: ```suggestion
 : A network device (usually located at the edge of the network) that makes forwarding decisions based on CATS information to steer traffic specific to a  service demand towards a corresponding yet selected service instance. The selection of a service instance relies upon a multi-metric CATS-based path computation.
```

Ingress CATS-Router:

>  : A CATS-Router that serves as a service access point for CATS clients. It steers service-specific traffic along a CATS-computed path that leads to an Egress CATS-Router that connects to the most suitable edge site that hots the service instance selected to satisfy the initial service demand.

Reviewer: ```suggestion
 : A  node that serves as a service access point for CATS clients. It steers service-specific traffic along a CATS-computed path that leads to an Egress CATS-Router that connects to the most suitable edge site that hots the service instance selected to satisfy the initial service demand.
```

Egress CATS-Router:

> : A CATS-Router that is located at the end of a CATS-cmputed path and which connects to a CATS-serviced site.

Reviewer: ```suggestion
: A node that is located at the end of a CATS-computed path and which connects to a CATS-serviced site.
```

CATS Service Metric Agent (C-SMA):

>  : An agent that is responsible for collecting service capabilities and status, and for reporting them to a CATS Path Selector (C-PS). Such agents may be hosted by CATS-Routers.

Reviewer: ```suggestion
 : An agent that is responsible for collecting service capabilities and status, and for reporting them to a CATS Path Selector (C-PS).```

CATS Network Metric Agent (C-NMA):

>   : An agent that is responsible for collecting network capabilities and status, and for reporting them to a C-PS. Such agents may be hosted by CATS-Routers.

Reviewer: ```suggestion
  : An agent that is responsible for collecting network capabilities and status, and for reporting them to a C-PS.

CATS Path Selector (C-PS):

>  : A computation logic that calculates and selects paths towards service locations and  instances and which accommodate the requirements of service demands. Such path computation engine takes into account the service and network status information.

Reviewer: ```suggestion
 : A computation logic that calculates and selects paths towards service locations and  instances and which accommodate the requirements of service demands. Such a path computation engine takes into account the service and network status information.
```

CATS Traffic Classifier (C-TC):

>  : The C-TC is responsible for determining which packets belong to a traffic flow for a particular service demand. It is also responsible for forwarding such packets along the C-PS-computed  path that leads to the relevant service instance.

Reviewer: ```suggestion
 : A functional entity that is responsible for determining which packets belong to a traffic flow for a particular service demand. It is also responsible for forwarding such packets along the C-PS-computed  path that leads to the relevant service instance.
```

# Framework and Components {#Framework-and-concepts}

## Assumptions

> CATS assumes that there are multiple service instances running on different edge nodes, and which provide a given service that is represented by a global service identifier (see {{cats-ids}}.

Reviewer: ```suggestion
CATS assumes that there are multiple service instances running on different edge nodes, and which provide a given service that is represented by a service identifier (see {{cats-ids}}. 
```

Note that, depending on design considerations and service requirements, per- service instance computing-related metrics or aggregated per-site computing-related metrics (and a combination thereof) can be used by a C-PS. Using aggregated per-site computing-related metrics appears as a privileged option scalability-wise, but relies on CATS-Routers that connect to various service instances to select the proper service instance.

In {{fig-cats-fw}}, the "underlay infrastructure" indicates an IP/MPLS network that is not necessarily CATS-aware. The P-CS-computed CATS routes will be distributed among the overlay CATS-Routers, and will not affect the underlay nodes. A CATS implementation may rely upon a control plane or a management plane  to distribute service metrics and network metrics - this document does not define a specific solution.

## Deployment Considerations

> This document does not make any assumption about how the various CATS functional elements are implmented and deployed. Whether a CATS deployment follows a fully distributed design or relies upon a mix of centralized (e.g., a C-PS) and distributed CATS functions (e.g., CATS traffic classifiers) is deployment-specific and may rflect the savoir-faire of the (CATS) service provider. Centralized designs where the computing-related metrics from the C-SMAs are collected by a (logically) centralized path computation logic (like a PCE) that also collects/defines network metrics may thus be adopted. In the latter case, the (logically) centralized CATS computation logic may thus process incoming service requests to compute and select paths let alone service instances. The outcomes of such computation may then b communicated to CATS traffic classifiers.

Reviewer: ```suggestion
This document does not make any assumption about how the various CATS functional elements are implmented and deployed. Whether a CATS deployment follows a fully distributed design or relies upon a mix of centralized (e.g., a C-PS) and distributed CATS functions (e.g., CATS traffic classifiers) is deployment-specific and may rflect the savoir-faire of the (CATS) service provider. Centralized designs where the computing-related metrics from the C-SMAs are collected by a (logically) centralized path computation logic (like a Path Computation Element (PCE) {{?RFC4655}}) that also collects network metrics may, thus, be adopted. In the latter case, the (logically) centralized CATS computation logic may process incoming service requests to compute and select paths let alone service instances. The outcomes of such a computation may then be communicated to CATS traffic classifiers (C-TCs).
```

# CATS Framework Workflow

The following subsections provide an overview of how the CATS workflow operates assuming a distributed CATS design.

## Service Announcement

A service is associated with a unique identifier called a CS-ID. A CS-ID may be a network identifier, such as an IP address. The mapping of CS-IDs to network identifiers may be learned through a name resolution service such as DNS.

## Metrics Distribution

As described above, a C-SMA collects both service-related capabilities and metrics, and associates them with a CS-ID that identifies the service. It may aggregate the metrics for multiple service instances, or maintain them separately or both. The C-SMA then advertises the CS-IDs along with the metrics to be received by all C-PSs in the network. The service-related metrics include computing-related metrics and potentially other service metrics like the number of end-users who access the service instance at any given time, their location and whether they are in motion or not.

> Computing metrics may change very frequently (see {{?I-D.yao-cats-ps-usecases}} for a discussion). How frequently such information is distributed is to be determined as part of the specification of any communication protocol (including routing protocols like OSPF) that may be used to distribute the information. Various options can be considered, such as interval-based updates, threshold-triggered updates, policy-based updates, etc.

Reviewer: ```suggestion
Computing metrics may change very frequently (see {{?I-D.yao-cats-ps-usecases}} for a discussion). How frequently such information is distributed is to be determined as part of the specification of any communication protocol (including routing protocols) that may be used to distribute the information. Various options can be considered, such as interval-based updates, threshold-triggered updates, policy-based updates, etc.
```

Additionally, the C-NMA collects network-related capabilities and metrics. These may be collected and distributed by existing routing protocols, although extensions to such protocols may be required to carry additional information such as link latency. The C-NMA distributes the network metrics to the C-PSes so that they can use the combination of service metrics and network metrics to determine the best CATS-Router to provide access to a service instance and invoke the compute function required by a service demand.

> Network metrics may also change over time. Dynamic routing protocols may take advantage of some information or capabilities (like Partial Route Computation of OSPFv3) to prevent the network from being flooded with state change information. C-NMA agents should also be configured or instructed like C-SMA agents to determine when and how often updates should be notified to the C-PSes.

Reviewer: ```suggestion
Network metrics may also change over time. Dynamic routing protocols may take advantage of some information or capabilities (e.g., Partial Route Computation (PRC) of OSPFv3 {{?RFC5340}}) to prevent the network from being flooded with state change information. C-NMA agents should also be configured or instructed like C-SMA agents to determine when and how often updates should be notified to the C-PSes.
```

{{fig-cats-example-overlay}} shows an example of how CATS metrics can be distributed. There is a client attached to the netowrk via CATS-Router 1. There are three instances of the service with CS-ID 1: two are located at Edge Site 2 attached via CATS-Router 2 and have CB-IDs 1 and 2; the third is located at Edge Site 3 attached via CATS-Router 3 and with CB-ID 3. There is also a second service with CS-ID 2 with only one instance located at Edge Site 2.

In this example, the C-SMA collocated with CATS-Router 2 distributes the service metrics for both service instances. Note that this information could be aggregated into a single advertisement, but in this case, the metrics for each instance are advertised separately. Similarly, the C-SMA agent located at Edge Site 2 advertises the service metrics for the two services hosted by Edge Site 2.

> The service metric advertisements are processed by the C-PS hosted by CATS-Router 1. The C-PS also processes network metric advertisements sent by the C-NMA agent. All metrics are used by the C-PS to compute and select the most relevant path that leads to the CATS-Router according to the initial  client's service demand, the service that is requested (CS-ID 1 or CS-ID 2), the state of the service instances as reported by the metrics, and the state of the network.

Reviewer: ```suggestion
The service metric advertisements are processed by the C-PS hosted by CATS-Router 1. The C-PS also processes network metric advertisements sent by the C-NMA agent. All metrics are used by the C-PS to compute and select the most relevant path that leads to the Egress CATS-Router according to the initial  client's service demand, the service that is requested (CS-ID 1 or CS-ID 2), the state of the service instances as reported by the metrics, and the state of the network.
```

~~~
{: #fig-cats-example-overlay title="Example CATS Metric Distribution"}

Figure 2 mainly describes a per-instance computing-related metric distribution. In the case of distributing aggregated per-site computing-related metrics, the per-instance CB-ID information will not be included in the advertisement. Instead, a per-site CB-ID may be used in case multiple sites are connected to the CATS router to explicitly indicate the site the aggregated metrics come from.

However, a CB-ID is not required if the edge site can support consistently service instance selection.

## Service Demand Processing

> The C-PS computes and selects paths that lead to CATS-Routers according to the service and network metrics that have been advertised. The C-PS may be collocated with a CATS-Router (as shown in {{fig-cats-example-overlay}}) or logically centralized.

Reviewer: ```suggestion
The C-PS computes and selects paths that lead to Egress CATS-Routers according to the service and network metrics that have been advertised. The C-PS may be collocated with a Ingress CATS-Router (as shown in {{fig-cats-example-overlay}}) or logically centralized.
```

This document does not specify any algorithm for path computation and selection purposes, but it is expected that a service demand or local policy may feed the C-PS computation logic with Objective Functions that provide some information about the path characteristics (e.g., in terms of maximum latency) and the selected service instance.

In the example of {{fig-cats-example-overlay}}, when the client sends a service demand to CATS-Router 1, the router solicits

> the C-PS to select a service instance hosted by an edge site that can be accessed through a particular CATS-Router. The C-PS also determines a path to the CATS-Router. This information is provided to CATS-Router 1 so that it can forward packets to their proper destination, as computed by the C-PS.

Reviewer: ```suggestion
the C-PS to select a service instance hosted by an edge site that can be accessed through a particular CATS-Router. The C-PS also determines a path to the Egress CATS-Router. This information is provided to Ingress CATS-Router (CATS-Router 1) so that it can forward packets to their proper destination, as computed by the C-PS.
```

> A service transaction consists of one or more service packets sent by the client to the CATS-Router it is connected to. The CATS-Router classifies incoming packets by soliciting the C-TC classifier, and forwards them to the C-PS-selected  CATS-Router. When these packets reach the CATS-Router, the outer header of the possible overlay encapsulation is removed and resulting packets are sent to the relevant service instance.

Reviewer: ```suggestion
A service transaction consists of one or more service packets sent by the client to an Ingress CATS-Router it is connected to. The Ingress CATS-Router classifies incoming packets by soliciting the C-TC classifier, encapsulates, and then forwards them to the C-PS-selected  Egress CATS-Router. When these packets reach the Egress CATS-Router, the outer header of the possible overlay encapsulation is removed and resulting packets are sent to the relevant service instance.
```

## Instance Affinity

> Instance affinity is a key feature of CATS. It means that packets that belong to a given flow associated with a given service should always be sent to the same CATS-Router which will forward them the same service instance. Furthermore, packets of a given flow should be forwarded along the same path to minimizie the risk of jeopardizing the integrity of the communication established between thee client and the service instance.

Reviewer: ```suggestion
Instance affinity is a key feature of CATS. It means that packets that belong to a given flow associated with a given service should always be sent to the same Egress CATS-Router which will forward them the same service instance. Furthermore, packets of a given flow should be forwarded along the same path to minimize the risk of jeopardizing the integrity of the communication established between thee client and the service instance.
```

More importantly, the means for identifying a flow for the purpose of ensuring instance affinity must be application-independent to avoid the need for service-specific instance affinity methods.

Specifically, service instance affinity information should be configurable on a per-service basis. For each service, the information can include the flow/packets identification type and means, affinity timeout value, etc.

This document does not define any mechanism for defining or enforcing service instance affinity.

# Security Considerations

The computing resource information changes over time very frequently, especially with the creation and termination of service instances. When such an information is carried in a routing protocol, too many updates may affect network stability. This issue could be exploited by an attacker (e.g., by spawning and deleting service instances very rapidly). CATS solutions must support guards against such misbehaviors. For example, these solutions should support aggregation techniques, dampening mechanisms, and threshold-triggered distribution updates.

> The information distributed by the C-SMA and C-NMA agents may be sensitive. Such information could indeed disclose intel about the network and the location of compute resources hosted in edge sites. This information may be used by an attacker to identify weak spots in an operator's network. Furthermore, such information may be modified by an attacker resulting in disrupted service delivery for the clients, up to and including misdirection of traffic to an attacker's service implementation. CATS solutions must support authentication and integrity-protection mechanisms between C-SMAs/C-NMAs and C-PSes, and between C-PSes and CATS-Routers. Also, C-SMA agents need to support a mechanism to authenticate the services for which they provide information to C-PS computation logics, among other CATS functions.

Reviewer: ```suggestion
The information distributed by the C-SMA and C-NMA agents may be sensitive. Such information could indeed disclose intel about the network and the location of compute resources hosted in edge sites. This information may be used by an attacker to identify weak spots in an operator's network. Furthermore, such information may be modified by an attacker resulting in disrupted service delivery for the clients, up to and including misdirection of traffic to an attacker's service implementation. CATS solutions must support authentication and integrity-protection mechanisms between C-SMAs/C-NMAs and C-PSes, and between C-PSes and Ingress CATS-Routers. Also, C-SMA agents need to support a mechanism to authenticate the services for which they provide information to C-PS computation logics, among other CATS functions.
```
